sycl/doc/extensions/proposed/sycl_ext_oneapi_free_function_kernels.asciidoc

```diff
@@ -0,0 +1,684 @@
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 8 specification.
+All references below to the "core SYCL specification" or to section numbers in
+the SYCL specification refer to that revision.
+
+This extension also depends on the following other SYCL extensions:
+
+* link:../proposed/sycl_ext_oneapi_free_function_queries.asciidoc[
+  sycl_ext_oneapi_free_function_queries]
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.
+Interfaces defined in this specification may not be implemented yet or may be
+in a preliminary state.
+The specification itself may also change in incompatible ways before it is
+finalized.
+*Shipping software products should not rely on APIs defined in this
+specification.*
+
+
+== Overview
+
+This extension introduces a new way to define a kernel as a simple C++
+function, where the kernel arguments are parameters to the function.
+This is different from standard SYCL kernels, where the kernel arguments are
+either captures of a lambda expression or member variables of a callable
+object.
+
+The primary motivation for this extension is the
+link:../proposed/sycl_ext_oneapi_kernel_compiler.asciidoc[
+sycl_ext_oneapi_kernel_compiler], which allows online compilation of a kernel
+from source code.
+These kernels must have a clean separation between host and device code, and
+they must have a defined order to their arguments because the application sets
+the argument values by their "index" via `handler::set_arg(index, value)`.
+Because variables captured by lambda expressions have no defined order, we need
+some other way to define kernel arguments for use with the
+sycl_ext_oneapi_kernel_compiler extension, and the
+sycl_ext_oneapi_free_function_kernels extension provides that mechanism.
+The "free function kernel" feature is a separate extension, though, because it
+can also be used independently from sycl_ext_oneapi_kernel_compiler.
+For example, some users may find it more familiar to define kernels as plain
+functions or they might have other reasons to prefer a clean separation between
+host and device code.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.
+An implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_FREE_FUNCTION_KERNELS`
+to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== Defining a free function kernel
+
+A free function kernel is a normal C++ function definition, where the function
```

Comment:
Just noting here that it opens up a possibility to have `constexpr` or `inline` kernel now. But I don't think that any of those should cause problems or complicate the implementation.

Reply:
Definition a kernel as `inline` is possibly interesting because it would allow the same kernel to be defined multiple times (in different translation units).  The linker will discard all but one definition, so all references to the kernel will get the same address.

I also can't think of any problem this will cause the implementation, but we should include a test for this case.